farm-boosted-rewards-subscriber/src/lib.rs

```diff
@@ -0,0 +1,162 @@
+        self.simple_lock_address().set(simple_lock_address);
+        self.lock_period().set(lock_period);
+
+        self.total_fees().set(UniquePayments::new());
+    }
+
+    #[endpoint(performAction)]
+    fn perform_action_endpoint(
+        &self,
+        service_index: usize,
+        users_to_claim: MultiValueEncoded<AddressId>,
+    ) -> OperationCompletionStatus {
+        require!(service_index <= 1, "invalid index");
+
+        let total_users = users_to_claim.len();
+        let mut args_vec = ManagedVec::new();
+        for _ in 0..total_users {
+            args_vec.push(AdditionalFarmData { dummy_data: 0 });
+        }
+
+        let current_epoch = self.blockchain().get_block_epoch();
+        let mut user_index = self.get_user_index(service_index, current_epoch);
+        let result = self.perform_service::<FarmClaimBoostedWrapper<Self>>(
+            total_users as u64 * (BUY_MEX_COST + LOCK_GAS_PER_USER),
+            service_index,
+            &mut user_index,
+            args_vec,
+        );
+
+        // do stuff with result
+
+        self.user_index().set(user_index);
+        self.last_global_action_epoch(service_index)
+            .set(current_epoch);
+
+        result.status
+    }
+}
+
+pub struct FarmClaimBoostedWrapper<
+    T: AllBaseTraits
+        + buy_mex::BuyMexModule
+        + claim_farm_boosted::ClaimFarmBoostedRewardsModule
+        + energy_query::EnergyQueryModule,
+> {
+    _phantom: PhantomData<T>,
+}
+
+impl<T> SubscriberContract for FarmClaimBoostedWrapper<T>
+where
+    T: AllBaseTraits
+        + buy_mex::BuyMexModule
+        + claim_farm_boosted::ClaimFarmBoostedRewardsModule
+        + energy_query::EnergyQueryModule,
+{
+    type SubSc = T;
+    type AdditionalDataType = AdditionalFarmData;
+
+    fn perform_action(
+        sc: &Self::SubSc,
+        user_address: ManagedAddress<<Self::SubSc as ContractBase>::Api>,
+        fee: EgldOrEsdtTokenPayment<<Self::SubSc as ContractBase>::Api>,
+        service_index: usize,
+        service_info: &ServiceInfo<<Self::SubSc as ContractBase>::Api>,
+        _additional_data: &<Self as SubscriberContract>::AdditionalDataType,
+    ) -> Result<InterpretedResult<<Self::SubSc as ContractBase>::Api>, ()> {
+        if service_index == 1 {
+            let user_energy = sc.get_energy_amount(&user_address);
+            let energy_threshold = sc.energy_threshold().get();
+            if user_energy < energy_threshold {
+                return Result::Err(());
+            }
+        }
+
+        let actions_percentage = if service_index == 0 {
+            sc.normal_user_percentage().get()
+        } else {
+            sc.premium_user_percentage().get()
+        };
+
+        let token_id = if fee.token_identifier.is_egld() {
+            // wrap egld
+            TokenIdentifier::from("PLACEHOLDER")
+        } else {
+            fee.token_identifier.unwrap_esdt()
+        };
+
+        sc.perform_mex_operations(
```

Comment:
we do not need to do the mex_operations every time we do a perform action for the user.

Taking subscription FEE and doing MEX operations is ONCE per MONTH.

ClaimingBoostedRewards is ONCE per WEEK.

Make a separate endpoint for taking payment from subscription contract. Also which function is doing the take payment ? :)